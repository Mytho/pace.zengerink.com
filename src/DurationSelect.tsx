import React from 'react'
import NumberSelect from './NumberSelect'

interface DurationSelectProps {
    value: number
}

interface DurationSelectState {}

interface Duration {
    hours: number
    minutes: number
    seconds: number
}

export default class DurationSelect extends React.Component<DurationSelectProps, DurationSelectState> {
    toDuration(seconds: number): Duration {
        let date = new Date(seconds * 1000);

        return {
            hours: date.getUTCHours(),
            minutes: date.getUTCMinutes(),
            seconds: date.getSeconds() 
        }
    }

    render () {
        const duration = this.toDuration(this.props.value);

        return (
            <div className="duration-select">
                <NumberSelect value={duration.hours} max="23" />
                <span>:</span>
                <NumberSelect value={duration.minutes} max="59" pad="2" />
                <span>:</span>
                <NumberSelect value={duration.seconds} max="59" pad="2" />
                <span>hh:mm:ss</span>
            </div>
        )
    }
}