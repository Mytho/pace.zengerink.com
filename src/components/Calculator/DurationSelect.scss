@import './Calculator.scss';

.duration-select {
    @include calculator-row;
}