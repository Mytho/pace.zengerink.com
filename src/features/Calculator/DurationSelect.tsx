import React from 'react'
import NumberSelect from './NumberSelect'
import './DurationSelect.scss'

export enum DurationFragment {
  Hours = 'hours',
  Minutes = 'minutes',
  Seconds = 'seconds',
}

interface DurationSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

interface Duration {
  [key: string]: number
}

export default class DurationSelect extends React.Component<DurationSelectProps, {}> {
  handleChange(prop: string, value: number) {
    const duration = this.toDuration(this.props.value)
    duration[prop] = value
    this.props.onChange(this.toSeconds(duration))
  }

  toSeconds(duration: Duration): number {
    const hours = duration[DurationFragment.Hours] * 3600
    const minutes = duration[DurationFragment.Minutes] * 60

    return hours + minutes + duration[DurationFragment.Seconds]
  }

  toDuration(seconds: number): Duration {
    const date = new Date(seconds * 1000)

    return {
      [DurationFragment.Hours]: date.getUTCHours(),
      [DurationFragment.Minutes]: date.getUTCMinutes(),
      [DurationFragment.Seconds]: date.getSeconds(),
    }
  }

  render() {
    const duration = this.toDuration(this.props.value)

    return (
      <div className="duration-select">
        <div className="duration-select__label">{this.props.label}</div>
        <div className="duration-select__input">
          <NumberSelect
            name={DurationFragment.Hours}
            value={duration[DurationFragment.Hours]}
            max="23"
            onChange={this.handleChange.bind(this, DurationFragment.Hours)}
          />
          <span className="duration-select__value-label">hrs</span>
          <NumberSelect
            name={DurationFragment.Minutes}
            value={duration[DurationFragment.Minutes]}
            max="59"
            pad="2"
            onChange={this.handleChange.bind(this, DurationFragment.Minutes)}
          />
          <span className="duration-select__value-label">min</span>
          <NumberSelect
            name={DurationFragment.Seconds}
            value={duration[DurationFragment.Seconds]}
            max="59"
            pad="2"
            onChange={this.handleChange.bind(this, DurationFragment.Seconds)}
          />
          <span className="duration-select__value-label">sec</span>
        </div>
      </div>
    )
  }
}
