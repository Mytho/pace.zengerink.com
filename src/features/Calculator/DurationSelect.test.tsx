import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import DurationSelect from './DurationSelect'

describe('render', () => {
  it('renders correctly', () => {
    const onChange = jest.fn()
    const wrapper = shallow(<DurationSelect label="Time" value={3661} onChange={onChange} />)
    const json = shallowToJson(wrapper)
    expect(json).toMatchSnapshot()
  })
})

describe('onChange', () => {
  const testHandleChange = (name: string, newValue: number, expectedArgument: number) => {
    const onChange = jest.fn()
    const wrapper = shallow(<DurationSelect label="Time" value={0} onChange={onChange} />)
    const select = wrapper.find(`[name="${name}"]`)
    expect(select.length).toBe(1)

    select.simulate('change', newValue)
    expect(onChange.mock.calls.length).toBe(1)
    expect(onChange.mock.calls[0][0]).toBe(expectedArgument)
  }

  it('is called when one of the selects changes', () => {
    testHandleChange('hours', 1, 3600)
    testHandleChange('minutes', 2, 120)
    testHandleChange('seconds', 3, 3)
  })
})
