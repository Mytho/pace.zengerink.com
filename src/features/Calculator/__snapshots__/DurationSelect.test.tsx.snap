// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<div
  className="duration-select"
>
  <div
    className="duration-select__label"
  >
    Time
  </div>
  <div
    className="duration-select__input"
  >
    <NumberSelect
      max="23"
      name="hours"
      onChange={[Function]}
      value={1}
    />
    <span
      className="duration-select__value-label"
    >
      hrs
    </span>
    <NumberSelect
      max="59"
      name="minutes"
      onChange={[Function]}
      pad="2"
      value={1}
    />
    <span
      className="duration-select__value-label"
    >
      min
    </span>
    <NumberSelect
      max="59"
      name="seconds"
      onChange={[Function]}
      pad="2"
      value={1}
    />
    <span
      className="duration-select__value-label"
    >
      sec
    </span>
  </div>
</div>
`;
